openhcl/underhill_mem/src/mapping.rs

```diff
@@ -599,4 +611,50 @@ unsafe impl GuestMemoryAccess for GuestMemoryMapping {
                 .map(|bitmap| bitmap.access_bitmap())
         }
     }
+
+    fn page_fault(
+        &self,
+        address: u64,
+        _len: usize,
+        _write: bool,
+        bitmap_failure: bool,
+    ) -> guestmem::PageFaultAction {
+        let err = {
+            if bitmap_failure {
+                if !self
+                    .valid_memory
+                    .as_ref()
+                    .expect("all backings with bitmaps should have a GuestValidMemory")
+                    .check_valid(address / PAGE_SIZE as u64)
+                {
+                    if self.iova_offset == Some(0) {
+                        guestmem::PageFaultError::new(
+                            guestmem::GuestMemoryErrorKind::NotPrivate,
+                            BitmapFailure::IncorrectHostVisibilityAccess,
+                        )
+                    } else {
+                        guestmem::PageFaultError::new(
+                            guestmem::GuestMemoryErrorKind::NotShared,
+                            BitmapFailure::IncorrectHostVisibilityAccess,
+                        )
+                    }
+                } else {
+                    // If there was a bitmap failure, and it wasn't the valid memory
+                    // bitmap, it must be the permissions bitmaps.
```

Comment:
You should probably double check the effective bitmap here. It's possible we're accessing guest memory during a shared/private transition, where the original check in the effective bitmap failed, but then the shared/private transition completed and now both the effective bitmap and the shared/private bitmaps show they we have access. In that case, we don't want to confuse this for a VTL protection violation and cause an exit to VTL1 for the wrong page.

(If the permissions check passes here, then you can probably just resolve the page fault as "retry" and ignore this race condition.)